junit-platform-runner/src/main/java/org/junit/platform/runner/JUnitPlatformTestTree.java

```diff
@@ -60,28 +67,64 @@ Description getDescription(TestIdentifier identifier) {
 		return this.descriptions.get(identifier);
 	}
 
-	private Description generateSuiteDescription(TestPlan testPlan, Class<?> testClass) {
-		Description suiteDescription = Description.createSuiteDescription(testClass.getName());
-		buildDescriptionTree(suiteDescription, testPlan);
+	void executeTests(RunNotifier notifier) {
+		JUnitPlatformRunnerListener listener = new JUnitPlatformRunnerListener(this, notifier);
+		this.testCollection.execute(listener);
+	}
+
+	void filter(Filter filter) throws NoTestsRemainException {
+		Set<TestIdentifier> filteredIdentifiers = getFilteredLeaves(filter);
+		if (filteredIdentifiers.isEmpty()) {
+			throw new NoTestsRemainException();
+		}
+
+		PostDiscoveryFilter postDiscoveryFilter = (TestDescriptor t) -> {
+			return FilterResult.includedIf(filteredIdentifiers.contains(TestIdentifier.from(t)));
+		};
+		testCollection.applyPostDiscoveryFilters(postDiscoveryFilter);
+
+		// Update our fields to match the testCollection
+		descriptions.clear();
+		suiteDescription = generateSuiteDescription(testCollection, suiteDescription.getDisplayName());
+	}
+
+	private Description generateSuiteDescription(TestCollection testCollection, String suiteDisplayName) {
+		Description suiteDescription = Description.createSuiteDescription(suiteDisplayName);
+		buildDescriptionTree(suiteDescription, testCollection);
 		return suiteDescription;
 	}
 
-	private void buildDescriptionTree(Description suiteDescription, TestPlan testPlan) {
-		testPlan.getRoots().forEach(testIdentifier -> buildDescription(testIdentifier, suiteDescription, testPlan));
+	private void buildDescriptionTree(Description suiteDescription, TestCollection testCollection) {
+		TestPlan testPlan = testCollection.testPlan();
+		testPlan.getRoots().stream().filter(testIdentifier -> !testPlan.getChildren(testIdentifier).isEmpty()).forEach(
+			testIdentifier -> buildDescription(testIdentifier, suiteDescription, testPlan));
 	}
 
 	void addDynamicDescription(TestIdentifier newIdentifier, String parentId) {
-		Description parent = getDescription(this.plan.getTestIdentifier(parentId));
-		this.plan.add(newIdentifier);
-		buildDescription(newIdentifier, parent, this.plan);
+		TestPlan plan = this.testCollection.testPlan();
+		TestIdentifier parentTestIdentifier = plan.getTestIdentifier(parentId);
+		Description parent = getDescription(parentTestIdentifier);
+
+		if (parent == null) {
```

Comment:
BTW, I'm not entirely sure how the current code works without this `if` block (see the comment on line 113 for more context)